internal/controller/apiextensions/definition/managed.go

```diff
@@ -0,0 +1,257 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+type managedResourceInformers struct {
+	log logging.Logger
+	mgr manager.Manager
+
+	mrInformersByCRDName map[schema.GroupVersionKind]mrInformer
+
+	gvkRoutedCache *controller.GVKRoutedCache
+
+	lock  sync.RWMutex
+	sinks map[string]func(ev runtimeevent.UpdateEvent) // by some uid
+	xrs   map[string]xr                                // by controller name
+}
+
+type mrInformer struct {
+	cache    cache.Cache
+	cancelFn context.CancelFunc
+}
+
+type xr struct {
+	gvk   schema.GroupVersionKind
+	cache cache.Cache
+}
+
+var _ source.Source = &managedResourceInformers{}
+
+// Start implements source.Source, i.e. starting managedResourceInformers as
+// source with h as the sink of update events.
+func (i *managedResourceInformers) Start(ctx context.Context, h handler.EventHandler, q workqueue.RateLimitingInterface, ps ...predicate.Predicate) error {
+	id := uuid.New().String()
+
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.sinks[id] = func(ev runtimeevent.UpdateEvent) {
+		for _, p := range ps {
+			if !p.Update(ev) {
+				return
+			}
+		}
+		h.Update(ctx, ev, q)
+	}
+
+	go func() {
+		<-ctx.Done()
+
+		i.lock.Lock()
+		defer i.lock.Unlock()
+		delete(i.sinks, id)
+	}()
+
+	return nil
+}
+
+// RegisterXR maps a controller name to a GVK.
+func (i *managedResourceInformers) RegisterXR(name string, ca cache.Cache, gvk schema.GroupVersionKind) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if i.xrs == nil {
+		i.xrs = make(map[string]xr)
+	}
+	i.xrs[name] = xr{
+		gvk:   gvk,
+		cache: ca,
+	}
+}
+
+// UnregisterXR removes a controller name to GVK mapping.
+func (i *managedResourceInformers) UnregisterXR(name string) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	delete(i.xrs, name)
+}
+
+func (i *managedResourceInformers) Watch(gvks ...schema.GroupVersionKind) {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	// start new informers
+	for _, gvk := range gvks {
+		if _, found := i.mrInformersByCRDName[gvk]; found {
+			continue
+		}
+
+		log := i.log.WithValues("gvk", gvk.String())
+
+		ctx, cancelFn := context.WithCancel(context.Background()) //nolint:govet // the outer context is closed by the caller
+		ca, err := cache.New(i.mgr.GetConfig(), cache.Options{})
+		if err != nil {
+			log.Debug("failed creating a cache", "error", err)
+			continue
+		}
+
+		u := kunstructured.Unstructured{}
+		u.SetGroupVersionKind(gvk)
+		inf, err := ca.GetInformer(ctx, &u)
+		if err != nil {
+			cancelFn()
+			log.Debug("failed getting informer", "error", err)
+			continue
+		}
+
+		if _, err := inf.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				old := oldObj.(client.Object)
+				new := newObj.(client.Object) //nolint:revive // this is fine
```

Comment:
Was the linter complaining about redefining `new`? If so, worth mentioning in the comment. (i.e. "Redefining new here is fine"). "This is fine" alone doesn't tell me much as a reader. 🙂

Reply:
renamed to `obj`.